ieee802154: update KW41Z driver to also support KW40Z

bogdan-davidoaia opened the pull request.

bogdan-davidoaia committed:
ext: mcux: import XCVR driver for KW40Z

Import the KW40Z XCVR driver from Kinetis SDK v1.3, which is needed
to setup the radio on the hexiwear_kw40z board.

Origin: NXP Kinetis KW40Z Connectivity Software
Maintained-by: External

Signed-off-by: Bogdan Davidoaia <bogdan.davidoaia@linaro.org>

bogdan-davidoaia committed:
drivers: ieee802154: kw41z: fix issue in RX sequence

The chip seems to require a wait time before reading packet
data after the RX sequence done interrupt, as imidiate reading
of it will indicate a packet size of zero and cause the RX
sequence done to be continually retriggered.

In addition, the extra clearing of the interrupt flags can be
removed, as this is already done at the beginning of the isr.

Jira: ZEP-2278
Signed-off-by: Bogdan Davidoaia <bogdan.davidoaia@linaro.org>

bogdan-davidoaia committed:
samples: net: ieee802154: add KW40Z prj conf file

The config file for KW40Z has the shell disabled, as available
SRAM memory is limitted to 16kB.

Signed-off-by: Bogdan Davidoaia <bogdan.davidoaia@linaro.org>

bogdan-davidoaia committed:
ext: mcux: import additional KW40Z header files

Import from Kinetis SDK v1.3 MKW40Z4_extension.h and fsl_bitaccess.h
which are needed to build the XCVR driver for KW40Z.

Origin: NXP Kinetis KW40Z Connectivity Software
Maintained-by: External

Signed-off-by: Bogdan Davidoaia <bogdan.davidoaia@linaro.org>

bogdan-davidoaia committed:
ext: mcux: update KW40Z files for building XCVR driver

Update KW40Z files for building XCVR driver by adding the required
includes, removing unneeded ones and replacing OSA_EnterCritical/
OSA_ExitCritical with OSA_InterruptDisable/OSA_InterruptEnable.

Signed-off-by: Bogdan Davidoaia <bogdan.davidoaia@linaro.org>

bogdan-davidoaia committed:
ext: mcux: add makefiles for building KW40Z XCVR driver

Add and update makefiles in order to build either the KW41Z or KW40Z
XCVR driver based on soc name.

Signed-off-by: Bogdan Davidoaia <bogdan.davidoaia@linaro.org>

bogdan-davidoaia committed:
ext: mcux: add minimal v2.2 interface for KW40Z XCVR driver

Add minimal interface for the KW40Z XCVR driver to match the one offered
in MCUX v2.2 by the KW41Z XCVR driver.

This is needed to reuse the KW41Z IEEE 802.15.4 driver on KW40Z.

Signed-off-by: Bogdan Davidoaia <bogdan.davidoaia@linaro.org>

bogdan-davidoaia committed:
samples: net: ieee802154: kw41z: reduce log level

Reduce log level to show only error messages.

Signed-off-by: Bogdan Davidoaia <bogdan.davidoaia@linaro.org>

tbursztyka commented on drivers/ieee802154/ieee802154_kw41z.c:
```diff
@@ -313,6 +323,9 @@ static inline void kw41z_rx(struct kw41z_context *kw41z, u8_t len)
 	net_pkt_frag_insert(pkt, frag);
 
+#if CONFIG_SOC_MKW41Z4
+	u16_t reg_val = 0;
```
looks like you could put this declaration at the beginning of the for loop. It's not super clean to declare a variable elsewhere than in the beginning of a code block

bogdan-davidoaia replied:
Done. I've declared and initialized in the for loop.

tbursztyka commented on drivers/ieee802154/ieee802154_kw41z.c:
```diff
@@ -322,6 +335,27 @@ static inline void kw41z_rx(struct kw41z_context *kw41z, u8_t len)
 		}
 	}
+#else /* CONFIG_SOC_MKW40Z4 */
+	u32_t reg_val = 0;
```
same here

tbursztyka reviewed (changes requested):
Weird that github takes the last review as the whole, I though it could do it commit by commit. Anyway, switching to request changes (a tiny one)

MaureenHelm commented on drivers/ieee802154/ieee802154_kw41z.c:
```diff
@@ -322,6 +335,27 @@ static inline void kw41z_rx(struct kw41z_context *kw41z, u8_t len)
+	for (i = 0; i < pkt_len; i++) {
+		switch (i % 4) {
+		case 0:
+			reg_val = *(((u32_t *)ZLL->PKT_BUFFER) + i/4);
```
I know you're following the pattern from KW41 above, but is there a reason you didn't implement it like this:
```
reg_val = ZLL->PKT_BUFFER[i/4];
```

bogdan-davidoaia replied:
Good point. Also adding a patch to change it for KW41Z.

bogdan-davidoaia committed:
drivers: ieee802154: kw41z: add support for KW40Z

Add support for KW40Z in the KW41Z IEEE 802.15.4 driver as both chips
expose almost the same radio registers.

Signed-off-by: Bogdan Davidoaia <bogdan.davidoaia@linaro.org>

bogdan-davidoaia committed:
drivers: ieee802154: kw41z: remove unnecessary cast for PKT_BUFFER_RX

The cast to u16_t in not needed as PKT_BUFFER_RX already is
of that type.

Signed-off-by: Bogdan Davidoaia <bogdan.davidoaia@linaro.org>

bogdan-davidoaia commented on ext/hal/nxp/mcux/middleware/wireless/framework_5.3.3/XCVR/MKW40Z4/Makefile:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+obj-y += ../../OSAbstraction/Source/fsl_os_abstraction_zephyr.o
```
Done

bogdan-davidoaia commented on ext/hal/nxp/mcux/middleware/wireless/framework_5.3.3/XCVR/Makefile:
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+MCUX_DEVICE = $(shell echo $(CONFIG_SOC) | tr '[:lower:]' '[:upper:]')
```
Done

MaureenHelm commented:
I see a lot of warnings around BBF_STEP_RECIP_def_c when I build this. I don't remember seeing them in your previous revision, but it's hard to see what's different between revisions of a PR. This is one thing I liked about gerrit that we don't have in GitHub.

bogdan-davidoaia commented:
The warnings were there before. They all come from the XCVR driver imported from KSDK (which isn't really in line with the Zephyr coding style any way). I didn't address them as I didn't want to do too many changes in the imported code.

Should I add a patch to fix the warnings?

jukkar commented:
This patch has been pending long time, @MaureenHelm do you wish that that @bogdan-davidoaia should fix the warnings?

MaureenHelm commented:
> This patch has been pending long time, @MaureenHelm do you wish that that @bogdan-davidoaia should fix the warnings?

No, let's leave them alone. I agree with @bogdan-davidoaia 's initial assessment that it would be a lot of changes to imported code.

jukkar commented:
@MaureenHelm Can you approve what is currently there so we can merge this PR, or is there still some changes needed?
